pkg/server/service.go

```diff
@@ -156,30 +162,56 @@ func NewCRIContainerdService(config options.Config) (CRIContainerdService, error
 		return nil, fmt.Errorf("failed to create stream server: %v", err)
 	}
 
-	c.eventMonitor = newEventMonitor(c)
+	c.eventMonitor = newEventMonitor(c.containerStore, c.sandboxStore)
 
-	// Create the grpc server and register runtime and image services.
+	// To avoid race condition between `Run` and `Stop`, still create grpc server
+	// although we may not use it. It's just a small in-memory data structure.
+	// TODO(random-liu): Get rid of the grpc server when completely switch
+	// to plugin mode.
 	c.server = grpc.NewServer()
-	instrumented := newInstrumentedService(c)
-	runtime.RegisterRuntimeServiceServer(c.server, instrumented)
-	runtime.RegisterImageServiceServer(c.server, instrumented)
-	api.RegisterCRIContainerdServiceServer(c.server, instrumented)
 
-	return newInstrumentedService(c), nil
+	return c, nil
+}
+
+// Register registers all required services onto a specific grpc server.
+// This is used by containerd cri plugin.
+func (c *criContainerdService) Register(s *grpc.Server) error {
+	instrumented := newInstrumentedService(c)
+	runtime.RegisterRuntimeServiceServer(s, instrumented)
+	runtime.RegisterImageServiceServer(s, instrumented)
+	api.RegisterCRIContainerdServiceServer(s, instrumented)
+	return nil
 }
 
-// Run starts the cri-containerd service.
-func (c *criContainerdService) Run() error {
+// Run starts the cri-containerd service. startGRPC specifies
+// whether to start grpc server in this function.
+func (c *criContainerdService) Run(startGRPC bool) error {
```

Comment:
todo remove "startGRPC = true" case when we no longer support cri-containerd stand alone mode

Reply:
Will do.

Reply:
Done.